File: sdk/apple/Sources/DesfireEV3/Offline.swift
import CDesfire
import Foundation

/** Application metadata authenticated by delegated-application issuer MAC generation. */
public struct DelegatedApplicationConfiguration: Sendable {
    public var applicationId: ApplicationId
    public var keySettings: UInt8
    public var numberOfKeys: UInt8
    public var slot: UInt16
    public var slotVersion: UInt8
    public var quotaLimit: UInt16
    public var isoFileIdentifiers: Bool
    public var keySettings3: UInt8?
    public var isoId: UInt16?
    public var dfName: Data
    public var keySet: DelegatedKeySetConfiguration?

    /** Construct explicit delegated fields without choosing card policy defaults. */
    public init(
        applicationId: ApplicationId, keySettings: UInt8, numberOfKeys: UInt8,
        slot: UInt16, slotVersion: UInt8, quotaLimit: UInt16,
        isoFileIdentifiers: Bool = false, keySettings3: UInt8? = nil,
        isoId: UInt16? = nil, dfName: Data = Data(),
        keySet: DelegatedKeySetConfiguration? = nil
    ) throws {
        guard dfName.count <= 16 else {
            throw DesfireError(.invalidArgument, "Delegated DF name exceeds sixteen bytes")
        }
        self.applicationId = applicationId
        self.keySettings = keySettings
        self.numberOfKeys = numberOfKeys
        self.slot = slot
        self.slotVersion = slotVersion
        self.quotaLimit = quotaLimit
        self.isoFileIdentifiers = isoFileIdentifiers
        self.keySettings3 = keySettings3
        self.isoId = isoId
        self.dfName = dfName
        self.keySet = keySet
    }
}

/** Optional EV3 key-set fields included in delegated-application creation. */
public struct DelegatedKeySetConfiguration: Sendable {
    public var activeVersion: UInt8
    public var numberOfSets: UInt8
    public var maximumKeySize: UInt8
    public var settings: UInt8

    /** Retain the four exact documented key-set bytes. */
    public init(
        activeVersion: UInt8, numberOfSets: UInt8, maximumKeySize: UInt8,
        settings: UInt8
    ) {
        self.activeVersion = activeVersion
        self.numberOfSets = numberOfSets
        self.maximumKeySize = maximumKeySize
        self.settings = settings
    }
}

/** Independently owned transaction MAC and ReaderID session keys. */
public struct TransactionMacKeys: Sendable {
    public let macKey: Aes128Key
    public let encryptionKey: Aes128Key

    /** Adopt two exact AES keys returned by verified native derivation. */
    init(macKey: Aes128Key, encryptionKey: Aes128Key) {
        self.macKey = macKey
        self.encryptionKey = encryptionKey
    }
}

/** Stateless offline AES utilities. None of these methods communicates with a card. */
public enum Offline {
    /** Resolve one scoped key through the same direct, derived, or provider workflow. */
    private static func resolve(_ source: KeySource, purpose: KeyPurpose) async throws -> Data {
        guard let keyNumber = KeyNumber(rawValue: 0) else {
            throw DesfireError(.internal, "Invalid built-in offline key selector")
        }
        return try await resolveKeySource(
            source, keyNumber: keyNumber, profile: nil, scope: .native, purpose: purpose)
    }

    /** Overwrite one temporary Swift-owned secret copy. */
    private static func wipe(_ value: inout Data) {
        _ = value.withUnsafeMutableBytes {
            bytes in bytes.initializeMemory(as: UInt8.self, repeating: 0)
        }
        value.removeAll(keepingCapacity: false)
    }

    /** Convert one C owned-buffer result and preserve native failure evidence. */
    private static func output(
        _ operation: (UnsafeMutablePointer<OpaquePointer?>, UnsafeMutablePointer<df_error>) -> Int32
    ) throws -> Data {
        var buffer: OpaquePointer?
        var error = df_error()
        defer { df_buffer_free(buffer) }
        try checkNative(operation(&buffer, &error), error)
        return try ownedBuffer(buffer)
    }

    /** Derive one AES-128 key with the built-in NXP AN10922 construction. */
    public static func deriveNxpAes128(
        masterKey: KeySource, diversification: Data
    ) async throws -> Aes128Key {
        try validateNativeIdentity()
        var key = try await resolve(masterKey, purpose: .offlineOperation)
        defer { wipe(&key) }
        var value = try withBinary(key) { keyPointer, keyCount in
            try withBinary(diversification) { inputPointer, inputCount in
                try output {
                    df_offline_derive_nxp_aes128(
                        keyPointer, keyCount, inputPointer, inputCount, $0, $1)
                }
            }
        }
        defer { wipe(&value) }
        return try Aes128Key(value)
    }

    /** Encrypt one application default key into the documented 32-byte delegated EncK. */
    public static func encryptDelegatedDefaultKeyAes(
        damEncryptionKey: KeySource, applicationDefaultKey: KeySource,
        applicationDefaultKeyVersion: UInt8
    ) async throws -> Data {
        try validateNativeIdentity()
        var dam = try await resolve(damEncryptionKey, purpose: .delegatedApplication)
        var application = try await resolve(
            applicationDefaultKey, purpose: .delegatedApplication)
        defer {
            wipe(&dam)
            wipe(&application)
        }
        return try withBinary(dam) { damPointer, damCount in
            try withBinary(application) { applicationPointer, applicationCount in
                try output {
                    df_offline_encrypt_delegated_default_key_aes(
                        damPointer, damCount, applicationPointer, applicationCount,
                        UInt32(applicationDefaultKeyVersion), $0, $1)
                }
            }
        }
    }

    /** Calculate the issuer MAC authorizing one delegated-application creation. */
    public static func calculateDelegatedApplicationMacAes(
        damMacKey: KeySource, configuration: DelegatedApplicationConfiguration,
        encryptedDefaultKey: Data
    ) async throws -> Data {
        try validateNativeIdentity()
        var key = try await resolve(damMacKey, purpose: .delegatedApplication)
        defer { wipe(&key) }
        return try withBinary(key) { keyPointer, keyCount in
            try withBinary(configuration.dfName) { namePointer, nameCount in
                try withBinary(encryptedDefaultKey) { encryptedPointer, encryptedCount in
                    var native = df_delegated_application_configuration_v1()
                    native.struct_size = UInt32(MemoryLayout.size(ofValue: native))
                    native.abi_version = 1
                    native.application_id = configuration.applicationId.rawValue
                    native.key_settings = UInt32(configuration.keySettings)
                    native.number_of_keys = UInt32(configuration.numberOfKeys)
                    native.slot = UInt32(configuration.slot)
                    native.slot_version = UInt32(configuration.slotVersion)
                    native.quota_limit = UInt32(configuration.quotaLimit)
                    native.iso_file_identifiers = configuration.isoFileIdentifiers ? 1 : 0
                    native.key_settings3 = configuration.keySettings3.map(Int32.init) ?? -1
                    native.iso_id = configuration.isoId.map(Int32.init) ?? -1
                    native.df_name = namePointer
                    native.df_name_size = nameCount
                    if let keySet = configuration.keySet {
                        native.has_key_sets = 1
                        native.active_key_set_version = UInt32(keySet.activeVersion)
                        native.number_of_key_sets = UInt32(keySet.numberOfSets)
                        native.maximum_key_size = UInt32(keySet.maximumKeySize)
                        native.key_set_settings = UInt32(keySet.settings)
                    }
                    return try output {
                        df_offline_calculate_delegated_application_mac_aes(
                            keyPointer, keyCount, &native, encryptedPointer, encryptedCount,
                            $0, $1)
                    }
                }
            }
        }
    }

    /** Calculate the issuer MAC authorizing deletion of one delegated application. */
    public static func calculateDelegatedApplicationDeleteMacAes(
        damMacKey: KeySource, applicationId: ApplicationId
    ) async throws -> Data {
        try validateNativeIdentity()
        var key = try await resolve(damMacKey, purpose: .delegatedApplication)
        defer { wipe(&key) }
        return try withBinary(key) { pointer, count in
            try output {
                df_offline_calculate_delegated_application_delete_mac_aes(
                    pointer, count, applicationId.rawValue, $0, $1)
            }
        }
    }

    /** Calculate the DAM MAC for explicit existing and replacement ISO DF names. */
    public static func calculateDelegatedConfigurationMacAes(
        damMacKey: KeySource, oldDfName: Data, newDfName: Data
    ) async throws -> Data {
        try validateNativeIdentity()
        var key = try await resolve(damMacKey, purpose: .delegatedApplication)
        defer { wipe(&key) }
        return try withBinary(key) { keyPointer, keyCount in
            try withBinary(oldDfName) { oldPointer, oldCount in
                try withBinary(newDfName) { newPointer, newCount in
                    try output {
                        df_offline_calculate_delegated_configuration_mac_aes(
                            keyPointer, keyCount, oldPointer, oldCount, newPointer, newCount,
                            $0, $1)
                    }
                }
            }
        }
    }

    /** Calculate the documented MIFARE Classic compatibility-license authorization MAC. */
    public static func calculateMfcLicenseMacAes(
        licenseMacKey: KeySource, license: Data, sectorSecrets: Data
    ) async throws -> Data {
        try validateNativeIdentity()
        var key = try await resolve(licenseMacKey, purpose: .offlineOperation)
        defer { wipe(&key) }
        return try withBinary(key) { keyPointer, keyCount in
            try withBinary(license) { licensePointer, licenseCount in
                try withBinary(sectorSecrets) { secretPointer, secretCount in
                    try output {
                        df_offline_calculate_mfc_license_mac_aes(
                            keyPointer, keyCount, licensePointer, licenseCount,
                            secretPointer, secretCount, $0, $1)
                    }
                }
            }
        }
    }

    /** Derive independent transaction MAC and ReaderID encryption session keys. */
    public static func deriveTransactionMacKeysAes(
        transactionKey: KeySource, uid: Data, committedCounter: UInt32
    ) async throws -> TransactionMacKeys {
        try validateNativeIdentity()
        var key = try await resolve(transactionKey, purpose: .transactionMac)
        defer { wipe(&key) }
        var value = try withBinary(key) { keyPointer, keyCount in
            try withBinary(uid) { uidPointer, uidCount in
                try output {
                    df_offline_derive_transaction_mac_keys_aes(
                        keyPointer, keyCount, committedCounter, uidPointer, uidCount, $0, $1)
                }
            }
        }
        defer { wipe(&value) }
        guard value.count == 32 else {
            throw DesfireError(.internal, "Native transaction derivation returned invalid length")
        }
        return try TransactionMacKeys(
            macKey: Aes128Key(value.prefix(16)),
            encryptionKey: Aes128Key(value.suffix(16)))
    }

    /** Calculate one TMV from an already-derived transaction MAC session key. */
    public static func calculateTransactionMacSessionAes(
        sessionMacKey: Aes128Key, transactionInput: Data
    ) throws -> Data {
        try validateNativeIdentity()
        var key = try sessionMacKey.snapshot()
        defer { wipe(&key) }
        return try withBinary(key) { keyPointer, keyCount in
            try withBinary(transactionInput) { inputPointer, inputCount in
                try output {
                    df_offline_calculate_transaction_mac_session_aes(
                        keyPointer, keyCount, inputPointer, inputCount, $0, $1)
                }
            }
        }
    }

    /** Calculate one TMV directly from the backend key and explicit transaction evidence. */
    public static func calculateTransactionMacAes(
        transactionKey: KeySource, uid: Data, committedCounter: UInt32,
        transactionInput: Data
    ) async throws -> Data {
        try validateNativeIdentity()
        var key = try await resolve(transactionKey, purpose: .transactionMac)
        defer { wipe(&key) }
        return try withBinary(key) { keyPointer, keyCount in
            try withBinary(uid) { uidPointer, uidCount in
                try withBinary(transactionInput) { inputPointer, inputCount in
                    try output {
                        df_offline_calculate_transaction_mac_aes(
                            keyPointer, keyCount, committedCounter, uidPointer, uidCount,
                            inputPointer, inputCount, $0, $1)
                    }
                }
            }
        }
    }

    /** Verify one TMV; callers must separately enforce committed-counter replay policy. */
    public static func verifyTransactionMacAes(
        transactionKey: KeySource, uid: Data, committedCounter: UInt32,
        transactionInput: Data, transactionMac: Data
    ) async throws -> Bool {
        try validateNativeIdentity()
        var key = try await resolve(transactionKey, purpose: .transactionMac)
        defer { wipe(&key) }
        return try withBinary(key) { keyPointer, keyCount in
            try withBinary(uid) { uidPointer, uidCount in
                try withBinary(transactionInput) { inputPointer, inputCount in
                    try withBinary(transactionMac) { macPointer, macCount in
                        var verified: UInt32 = 0
                        var error = df_error()
                        try checkNative(
                            df_offline_verify_transaction_mac_aes(
                                keyPointer, keyCount, committedCounter, uidPointer, uidCount,
                                inputPointer, inputCount, macPointer, macCount, &verified, &error),
                            error)
                        return verified != 0
                    }
                }
            }
        }
    }

    /** Decrypt one authenticated EncTMRI with an already-derived transaction encryption key. */
    public static func decryptTransactionReaderIdAes(
        sessionEncryptionKey: Aes128Key, encryptedReaderId: Data
    ) throws -> Data {
        try validateNativeIdentity()
        var key = try sessionEncryptionKey.snapshot()
        defer { wipe(&key) }
        return try withBinary(key) { keyPointer, keyCount in
            try withBinary(encryptedReaderId) { encryptedPointer, encryptedCount in
                try output {
                    df_offline_decrypt_transaction_reader_id_aes(
                        keyPointer, keyCount, encryptedPointer, encryptedCount, $0, $1)
                }
            }
        }
    }

    /** Verify one raw-UID secp224r1 originality signature with an explicit trust anchor. */
    public static func verifyOriginalityUidSignature(
        publicKey: Data, uid: Data, signature: Data
    ) throws -> Bool {
        try validateNativeIdentity()
        return try withBinary(publicKey) { publicKeyPointer, publicKeyCount in
            try withBinary(uid) { uidPointer, uidCount in
                try withBinary(signature) { signaturePointer, signatureCount in
                    var verified: UInt32 = 0
                    var error = df_error()
                    try checkNative(
                        df_offline_verify_originality_uid_signature(
                            "secp224r1", publicKeyPointer, publicKeyCount, uidPointer, uidCount,
                            signaturePointer, signatureCount, &verified, &error), error)
                    return verified != 0
                }
            }
        }
    }
}

File: sdk/kotlin/src/main/kotlin/com/desfire/ev3/offline/Offline.kt
package com.desfire.ev3.offline

import com.desfire.ev3.KeyScope
import com.desfire.ev3.KeySource
import com.desfire.ev3.KeyNumber
import com.desfire.ev3.DelegatedApplication

/** Derive one AES-128 key using the documented NXP AN10922 construction. */
public fun deriveNxpAes128(masterKey: ByteArray, diversification: ByteArray): ByteArray {
    require(masterKey.size == 16) { "An AES-128 master key must contain sixteen bytes" }
    require(diversification.size in 1..31) {
        "Diversification input must contain one through 31 bytes"
    }
    return OfflineNative.deriveDirect(masterKey, diversification)
}

/** Resolve one provider master key and derive it using the provider diversification bytes. */
public fun deriveNxpAes128(
    masterKey: KeySource.Provider,
    keyNumber: KeyNumber = KeyNumber(0),
): ByteArray {
    require(masterKey.diversificationBytes.size in 1..31) {
        "Provider diversification input must contain one through 31 bytes"
    }
    return OfflineNative.deriveProvider(
        masterKey.bridge, KeyScope.NATIVE.code, keyNumber.value, masterKey.referenceBytes,
        masterKey.diversificationBytes, masterKey.contextBytes,
        masterKey.applicationId?.value ?: -1, masterKey.keySet?.value ?: -1,
    )
}

/** Calculate the documented eight-byte AES transaction MAC without card I/O. */
public fun calculateTransactionMacAes(
    transactionMacKey: ByteArray,
    transactionCounter: Long,
    uid: ByteArray,
    transactionInput: ByteArray,
): ByteArray {
    require(transactionMacKey.size == 16) { "An AES-128 transaction key must contain sixteen bytes" }
    require(transactionCounter in 0..0xFFFF_FFFFL) { "Transaction counter must fit uint32" }
    return OfflineNative.transactionMacDirect(
        transactionMacKey, transactionCounter, uid, transactionInput,
    )
}

/** Resolve one provider transaction key and calculate its eight-byte transaction MAC. */
public fun calculateTransactionMacAes(
    transactionMacKey: KeySource.Provider,
    keyNumber: KeyNumber = KeyNumber(0),
    transactionCounter: Long,
    uid: ByteArray,
    transactionInput: ByteArray,
): ByteArray {
    require(transactionCounter in 0..0xFFFF_FFFFL) { "Transaction counter must fit uint32" }
    return OfflineNative.transactionMacProvider(
        transactionMacKey.bridge, KeyScope.NATIVE.code, keyNumber.value,
        transactionMacKey.referenceBytes, transactionMacKey.diversificationBytes,
        transactionMacKey.contextBytes, transactionMacKey.applicationId?.value ?: -1,
        transactionMacKey.keySet?.value ?: -1, transactionCounter, uid, transactionInput,
    )
}

/** Verify a DESFire UID originality signature using the documented secp224r1 construction. */
public fun verifyOriginalityUidSignature(
    publicKey: ByteArray,
    uid: ByteArray,
    signature: ByteArray,
): Boolean = OfflineNative.verifyOriginality(publicKey, uid, signature)

/** Encrypt a delegated application default key into its exact 32-byte EncK record. */
public fun encryptDelegatedDefaultKeyAes(
    damEncryptionKey: ByteArray,
    applicationDefaultKey: ByteArray,
    applicationDefaultKeyVersion: Int,
): ByteArray {
    requireAesKey(damEncryptionKey)
    requireAesKey(applicationDefaultKey)
    require(applicationDefaultKeyVersion in 0..0xFF)
    return OfflineNative.invoke(0, longArrayOf(applicationDefaultKeyVersion.toLong()),
        arrayOf(damEncryptionKey, applicationDefaultKey))
}

/** Resolve DAMEncKey and encrypt a delegated application default key. */
public fun encryptDelegatedDefaultKeyAes(
    damEncryptionKey: KeySource.Provider,
    applicationDefaultKey: ByteArray,
    applicationDefaultKeyVersion: Int,
    providerKeyNumber: KeyNumber = KeyNumber(0),
): ByteArray {
    requireAesKey(applicationDefaultKey)
    require(applicationDefaultKeyVersion in 0..0xFF)
    return providerInvoke(damEncryptionKey, providerKeyNumber, 0,
        longArrayOf(applicationDefaultKeyVersion.toLong()), arrayOf(applicationDefaultKey))
}

/** Calculate the issuer MAC for one delegated application configuration and EncK. */
public fun calculateDelegatedApplicationMacAes(
    damMacKey: ByteArray,
    configuration: DelegatedApplication,
    encryptedDefaultKey: ByteArray,
): ByteArray {
    requireAesKey(damMacKey)
    return OfflineNative.invoke(1, configurationNumbers(configuration),
        arrayOf(damMacKey, configuration.dfNameBytes, encryptedDefaultKey))
}

/** Resolve DAMMACKey and calculate the issuer MAC for one delegated application. */
public fun calculateDelegatedApplicationMacAes(
    damMacKey: KeySource.Provider,
    configuration: DelegatedApplication,
    encryptedDefaultKey: ByteArray,
    providerKeyNumber: KeyNumber = KeyNumber(0),
): ByteArray = providerInvoke(damMacKey, providerKeyNumber, 1,
    configurationNumbers(configuration), arrayOf(configuration.dfNameBytes, encryptedDefaultKey))

/** Calculate the issuer MAC authorizing deletion of one delegated application. */
public fun calculateDelegatedApplicationDeleteMacAes(
    damMacKey: ByteArray,
    applicationId: Int,
): ByteArray {
    requireAesKey(damMacKey)
    require(applicationId in 1..0xFFFFFF)
    return OfflineNative.invoke(2, longArrayOf(applicationId.toLong()), arrayOf(damMacKey))
}

/** Resolve DAMMACKey and calculate the delegated application deletion MAC. */
public fun calculateDelegatedApplicationDeleteMacAes(
    damMacKey: KeySource.Provider,
    applicationId: Int,
    providerKeyNumber: KeyNumber = KeyNumber(0),
): ByteArray {
    require(applicationId in 1..0xFFFFFF)
    return providerInvoke(damMacKey, providerKeyNumber, 2,
        longArrayOf(applicationId.toLong()), arrayOf())
}

/** Calculate the issuer MAC over an old and replacement delegated DF name. */
public fun calculateDelegatedConfigurationMacAes(
    damMacKey: ByteArray,
    oldDfName: ByteArray,
    newDfName: ByteArray,
): ByteArray {
    requireAesKey(damMacKey)
    return OfflineNative.invoke(3, longArrayOf(), arrayOf(damMacKey, oldDfName, newDfName))
}

/** Resolve DAMMACKey and calculate the delegated DF-name configuration MAC. */
public fun calculateDelegatedConfigurationMacAes(
    damMacKey: KeySource.Provider,
    oldDfName: ByteArray,
    newDfName: ByteArray,
    providerKeyNumber: KeyNumber = KeyNumber(0),
): ByteArray = providerInvoke(damMacKey, providerKeyNumber, 3, longArrayOf(),
    arrayOf(oldDfName, newDfName))

/** Calculate the eight-byte AES MAC for one complete MIFARE Classic license record. */
public fun calculateMifareClassicLicenseMacAes(
    licenseMacKey: ByteArray,
    license: ByteArray,
    sectorSecrets: ByteArray,
): ByteArray {
    requireAesKey(licenseMacKey)
    return OfflineNative.invoke(4, longArrayOf(),
        arrayOf(licenseMacKey, license, sectorSecrets))
}

/** Resolve MFCLicenseMACKey and calculate a MIFARE Classic license MAC. */
public fun calculateMifareClassicLicenseMacAes(
    licenseMacKey: KeySource.Provider,
    license: ByteArray,
    sectorSecrets: ByteArray,
    providerKeyNumber: KeyNumber = KeyNumber(0),
): ByteArray = providerInvoke(licenseMacKey, providerKeyNumber, 4, longArrayOf(),
    arrayOf(license, sectorSecrets))

/** Derive SesTMMACKey followed by SesTMENCKey into one 32-byte result. */
public fun deriveTransactionMacKeysAes(
    transactionKey: ByteArray,
    transactionCounter: Long,
    uid: ByteArray,
): ByteArray {
    requireAesKey(transactionKey)
    requireCounter(transactionCounter)
    return OfflineNative.invoke(5, longArrayOf(transactionCounter), arrayOf(transactionKey, uid))
}

/** Resolve AppTransactionMACKey and derive both transaction session keys. */
public fun deriveTransactionMacKeysAes(
    transactionKey: KeySource.Provider,
    transactionCounter: Long,
    uid: ByteArray,
    providerKeyNumber: KeyNumber = KeyNumber(0),
): ByteArray {
    requireCounter(transactionCounter)
    return providerInvoke(transactionKey, providerKeyNumber, 5,
        longArrayOf(transactionCounter), arrayOf(uid))
}

/** Calculate an eight-byte TMV from one already-derived SesTMMACKey. */
public fun calculateTransactionMacSessionAes(
    sessionMacKey: ByteArray,
    transactionInput: ByteArray,
): ByteArray {
    requireAesKey(sessionMacKey)
    return OfflineNative.invoke(6, longArrayOf(), arrayOf(sessionMacKey, transactionInput))
}

/** Verify a transaction MAC from the backend key and exact committed inputs. */
public fun verifyTransactionMacAes(
    transactionKey: ByteArray,
    transactionCounter: Long,
    uid: ByteArray,
    transactionInput: ByteArray,
    transactionMac: ByteArray,
): Boolean {
    requireAesKey(transactionKey)
    requireCounter(transactionCounter)
    return booleanResult(OfflineNative.invoke(7, longArrayOf(transactionCounter),
        arrayOf(transactionKey, uid, transactionInput, transactionMac)))
}

/** Resolve AppTransactionMACKey and verify a transaction MAC. */
public fun verifyTransactionMacAes(
    transactionKey: KeySource.Provider,
    transactionCounter: Long,
    uid: ByteArray,
    transactionInput: ByteArray,
    transactionMac: ByteArray,
    providerKeyNumber: KeyNumber = KeyNumber(0),
): Boolean {
    requireCounter(transactionCounter)
    return booleanResult(providerInvoke(transactionKey, providerKeyNumber, 7,
        longArrayOf(transactionCounter), arrayOf(uid, transactionInput, transactionMac)))
}

/** Decrypt one exact 16-byte EncTMRI with an already-derived SesTMENCKey. */
public fun decryptTransactionReaderIdAes(
    sessionEncryptionKey: ByteArray,
    encryptedReaderId: ByteArray,
): ByteArray {
    requireAesKey(sessionEncryptionKey)
    require(encryptedReaderId.size == 16) { "Encrypted transaction reader ID must be 16 bytes" }
    return OfflineNative.invoke(8, longArrayOf(), arrayOf(sessionEncryptionKey, encryptedReaderId))
}

/** Encode the exact versioned delegated-application fields without native pointers. */
private fun configurationNumbers(configuration: DelegatedApplication): LongArray {
    val keySets = configuration.keySets
    return longArrayOf(
        configuration.applicationId.value.toLong(), configuration.keySettings.toLong(),
        configuration.numberOfKeys.toLong(), configuration.slot.toLong(),
        configuration.slotVersion.toLong(), configuration.quotaLimit.toLong(),
        if (configuration.isoFileIdentifiers) 1 else 0,
        configuration.keySettings3?.toLong() ?: -1,
        configuration.isoFileIdentifier?.value?.toLong() ?: -1,
        if (keySets != null) 1 else 0, keySets?.activeVersion?.toLong() ?: 0,
        keySets?.count?.toLong() ?: 0, keySets?.maximumKeySize?.toLong() ?: 0,
        keySets?.settings?.toLong() ?: 0,
    )
}

/** Invoke one provider helper with copied routing metadata. */
private fun providerInvoke(source: KeySource.Provider, keyNumber: KeyNumber, operation: Int,
    numbers: LongArray, data: Array<ByteArray>): ByteArray {
    return OfflineNative.invokeProvider(
        operation, source.bridge, KeyScope.NATIVE.code, keyNumber.value, source.referenceBytes,
        source.diversificationBytes, source.contextBytes, source.applicationId?.value ?: -1,
        source.keySet?.value ?: -1, numbers, data,
    )
}

/** Require one direct AES-128 key. */
private fun requireAesKey(key: ByteArray) {
    require(key.size == 16) { "An AES-128 key must contain sixteen bytes" }
}

/** Require an unsigned transaction counter. */
private fun requireCounter(counter: Long) {
    require(counter in 0..0xFFFF_FFFFL) { "Transaction counter must fit uint32" }
}

/** Decode the private one-byte JNI boolean representation. */
private fun booleanResult(value: ByteArray): Boolean {
    check(value.size == 1 && value[0].toInt() in 0..1) { "Malformed native boolean result" }
    return value[0].toInt() != 0
}

/** Private stateless JNI entry points; every returned array is independently owned. */
internal object OfflineNative {
    init { com.desfire.ev3.NativeRuntime.ensureLoaded() }

    @JvmStatic external fun deriveDirect(masterKey: ByteArray, diversification: ByteArray): ByteArray

    @JvmStatic external fun deriveProvider(
        provider: com.desfire.ev3.ProviderBridge,
        scope: Int,
        keyNumber: Int,
        reference: ByteArray,
        diversification: ByteArray,
        userContext: ByteArray,
        applicationId: Int,
        keySet: Int,
    ): ByteArray

    @JvmStatic external fun transactionMacDirect(
        key: ByteArray,
        transactionCounter: Long,
        uid: ByteArray,
        transactionInput: ByteArray,
    ): ByteArray

    @JvmStatic external fun transactionMacProvider(
        provider: com.desfire.ev3.ProviderBridge,
        scope: Int,
        keyNumber: Int,
        reference: ByteArray,
        diversification: ByteArray,
        userContext: ByteArray,
        applicationId: Int,
        keySet: Int,
        transactionCounter: Long,
        uid: ByteArray,
        transactionInput: ByteArray,
    ): ByteArray

    @JvmStatic external fun verifyOriginality(
        publicKey: ByteArray,
        uid: ByteArray,
        signature: ByteArray,
    ): Boolean

    @JvmStatic external fun invoke(
        operation: Int,
        numbers: LongArray,
        data: Array<ByteArray>,
    ): ByteArray

    @JvmStatic external fun invokeProvider(
        operation: Int,
        provider: com.desfire.ev3.ProviderBridge,
        scope: Int,
        keyNumber: Int,
        reference: ByteArray,
        diversification: ByteArray,
        userContext: ByteArray,
        applicationId: Int,
        keySet: Int,
        numbers: LongArray,
        data: Array<ByteArray>,
    ): ByteArray
}
